### DISCUSSION TITLE
Null Reference Exception with teardown & syncfusion tab view

### DISCUSSION THREAD

**Discussion by ajay-mundi** [AdamEssenmacher/MemoryToolkit.Maui#8] [Q&A]
Hi There,

First of all thanks for creating this. I am really excited to use this inside my app.

It seems like pretty much every single element in my views leaks. Is this really possible? If so wow. 

I have a shell app, the issue i seem to be running into is when I have tear down behaviour set to true in one of the root shell/flyout pages. Once i return to it the next time it is gone. Im guessing that because of the tear down. But is there any way around this or should I just not set teardown on these root shell/flyout pages? 

Also seems to be having an issue with pages containing syncfusion tab views, when I navigate backward I get a null reference error with the following exception.

`Syncfusion.Maui.TabView.TabViewMaterialVisualStyle.OnParentSet()
Microsoft.Maui.Controls.Element.SetParent(Element value)
Microsoft.Maui.Controls.Element.set_Parent(Element value)
MemoryToolkit.Maui.Utilities.<TearDown>g__TearDownImpl|2_0(IVisualTreeElement vte, Boolean isRoot)
MemoryToolkit.Maui.Utilities.<TearDown>g__TearDownImpl|2_0(IVisualTreeElement vte, Boolean isRoot)
MemoryToolkit.Maui.Utilities.<TearDown>g__TearDownImpl|2_0(IVisualTreeElement vte, Boolean isRoot)
MemoryToolkit.Maui.Utilities.<TearDown>g__TearDownImpl|2_0(IVisualTreeElement vte, Boolean isRoot)
MemoryToolkit.Maui.Utilities.<TearDown>g__TearDownImpl|2_0(IVisualTreeElement vte, Boolean isRoot)
MemoryToolkit.Maui.Utilities.<TearDown>g__TearDownImpl|2_0(IVisualTreeElement vte, Boolean isRoot)
MemoryToolkit.Maui.Utilities.<TearDown>g__TearDownImpl|2_0(IVisualTreeElement vte, Boolean isRoot)
MemoryToolkit.Maui.Utilities.<TearDown>g__TearDownImpl|2_0(IVisualTreeElement vte, Boolean isRoot)
MemoryToolkit.Maui.Utilities.<TearDown>g__TearDownImpl|2_0(IVisualTreeElement vte, Boolean isRoot)
MemoryToolkit.Maui.Utilities.<TearDown>g__TearDownImpl|2_0(IVisualTreeElement vte, Boolean isRoot)
MemoryToolkit.Maui.Utilities.TearDown(IVisualTreeElement vte)
MemoryToolkit.Maui.TearDownBehavior.OnVisualElementUnloaded(Object sender, EventArgs e)
System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
NSAsyncSynchronizationContextDispatcher.Apply () /Users/USER/azdo/_work/1/s/xamarin-macios/src/Foundation/NSAction.cs:176
Runtime.ThrowException (IntPtr gchandle) /Users/USER/azdo/_work/1/s/xamarin-macios/src/ObjCRuntime/Runtime.cs:2594
UIApplication.UIApplicationMain (Int32 argc, String[] argv, IntPtr principalClassName, IntPtr delegateClassName) /Users/USER/azdo/_work/1/s/xamarin-macios/src/UIKit/UIApplication.cs:60
UIApplication.Main (String[] args, Type principalClass, Type delegateClass) /Users/USER/azdo/_work/1/s/xamarin-macios/src/UIKit/UIApplication.cs:94
Program.Main (String[] args) /Users/USER/MauiApp/BTMobileApp/Platforms/iOS/Program.cs:14`

**Comment by AdamEssenmacher** [comment]
> It seems like pretty much every single element in my views leaks. Is this really possible? If so wow.

When leaks occur, this is actually virtually **guaranteed**--which is why this is such a big problem. Basically, all views eventually reference their parent view through the `.Parent` property, and parents reference their children for obvious reasons. So all views in a page indirectly reference each other. So, when one leaks... they all leak.

This toolkit can help you isolate actual 'root cause' leaky controls (don't assume there's just one on any given page!)

That said, it's _possible_ you're getting some false positives here. The behaviors make some guesses about when it thinks you actually are 'done with' a page/view. You can check for false positives at the page level by adding a quick `Debug.WriteLine("finalized!")` in the finalizer of the page you expect to be collected. If your finalizer runs, that means the page is getting GC'ed. If the toolkit reports this situation as a leak, then it's a false positive and you should let me know so we can fix it :)

As I mention in the readme, if `TearDown` runs before you want it to, it _will_ bork your app because it destroys views. I personally don't use Shell, so you could be using a configuration I didn't think to support.

> But is there any way around this or should I just not set teardown on these root shell/flyout pages?

If you have root pages that you expect to live through the lifetime of your app, then there's no benefit to adding the behaviors from the toolkit. The behaviors should be added to individual pages or views you expect to be GC'ed.
